cloudinit/sources/helpers/vultr.py

```diff
@@ -0,0 +1,254 @@
+# Author: Eric Benner <ebenner@vultr.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+import json
+import copy
+
+from cloudinit import log as log
+from cloudinit import url_helper
+from cloudinit import dmi
+from cloudinit import util
+from cloudinit import net
+from cloudinit.net.dhcp import EphemeralDHCPv4, NoDHCPLeaseError
+from functools import lru_cache
+
+# Get LOG
+LOG = log.getLogger(__name__)
+
+
+@lru_cache()
+def get_metadata(url, timeout, retries, sec_between):
+    # Bring up interface
+    try:
+        with EphemeralDHCPv4(connectivity_url=url):
+            # Fetch the metadata
+            v1 = read_metadata(url, timeout, retries, sec_between)
+    except (NoDHCPLeaseError) as exc:
+        LOG.error("Bailing, DHCP Exception: %s", exc)
+        raise
+
+    v1_json = json.loads(v1)
+    metadata = v1_json
+
+    # This comes through as a string but is JSON, make a dict
+    metadata['vendor-config'] = json.loads(metadata['vendor-config'])
+
+    return metadata
+
+
+# Read the system information from SMBIOS
+def get_sysinfo():
+    return {
+        'manufacturer': dmi.read_dmi_data("system-manufacturer"),
+        'subid': dmi.read_dmi_data("system-serial-number")
+    }
+
+
+# Assumes is Vultr is already checked
+def is_baremetal():
+    if get_sysinfo()['manufacturer'] != "Vultr":
+        return True
+    return False
+
+
+# Confirm is Vultr
+def is_vultr():
+    # VC2, VDC, and HFC use DMI
+    sysinfo = get_sysinfo()
+
+    if sysinfo['manufacturer'] == "Vultr":
+        return True
+
+    # Baremetal requires a kernel parameter
+    if "vultr" in util.get_cmdline():
+        return True
+
+    return False
+
+
+# Read Metadata endpoint
+def read_metadata(url, timeout, retries, sec_between):
+    url = "%s/v1.json" % url
+    response = url_helper.readurl(url,
+                                  timeout=timeout,
+                                  retries=retries,
+                                  headers={'Metadata-Token': 'vultr'},
+                                  sec_between=sec_between)
+
+    if not response.ok():
+        raise RuntimeError("Failed to connect to %s: Code: %s" %
+                           url, response.code)
+
+    return response.contents.decode()
+
+
+# Wrapped for caching
+@lru_cache()
+def get_interface_map():
+    return net.get_interfaces_by_mac()
+
+
+# Convert macs to nics
+def get_interface_name(mac):
+    macs_to_nic = get_interface_map()
+
+    if mac not in macs_to_nic:
+        return None
+
+    return macs_to_nic.get(mac)
+
+
+# Generate network configs
+def generate_network_config(md):
+    network = {
```

Comment:
**strongly suggest changing**

I think i'd rather this just take 'interfaces'... as that is actually all it uses.

Ie, its just kind of strange to have a function take a dictionary and then use only one value in the dictionary.

```python
func foo(mydata):
    bar = mydata['what-i-wanted']
    ... use bar here but no other part of 'mydata' and do some stuff...
```

Rather just make the caller call `foo(md['what-i-wanted'])`

Reply:
> **strongly suggest changing**
> 
> I think i'd rather this just take 'interfaces'... as that is actually all it uses.
> 
> Ie, its just kind of strange to have a function take a dictionary and then use only one value in the dictionary.
> 
> ```python
> func foo(mydata):
>     bar = mydata['what-i-wanted']
>     ... use bar here but no other part of 'mydata' and do some stuff...
> ```
> 
> Rather just make the caller call `foo(md['what-i-wanted'])`

you did this, thanks.
My only nitpick now is the variable name 'md'.... its a bit confusing.  change it to 'interfaces' ?

Reply:
thanks.